### PR TITLE
client: use vectors for context lists

### PR DESCRIPTION
**Pull request by batrick**
To avoid an allocation for each Context and improved cache locality.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by batrick** [comment]
jenkins test make check arm64

File: src/client/Client.cc [hunk 1]
```diff
@@ -4265,7 +4265,7 @@ void Client::signal_cond_list(list<ceph::condition_variable*>& ls)
   }
 }
 
-void Client::wait_on_context_list(list<Context*>& ls)
+void Client::wait_on_context_list(std::vector<Context*>& ls)
```
Comment (dparmar18) [inline src/client/Client.cc:4268]: We have already brought `std::vector` into the current scope at line 159 `using std::vector;`

File: src/client/Client.cc [hunk 2]
```diff
@@ -4276,30 +4276,14 @@ void Client::wait_on_context_list(list<Context*>& ls)
...
 
   // New items may have been added to the pending list, move them onto the
   // waitfor_caps list
-  while (!in->waitfor_caps_pending.empty()) {
-    Context *ctx = in->waitfor_caps_pending.front();
-
-    in->waitfor_caps_pending.pop_front();
-    in->waitfor_caps.push_back(ctx);
-  }
+  std::swap(in->waitfor_caps, in->waitfor_caps_pending);
```
Comment (dparmar18) [inline src/client/Client.cc:4286]: the iteration above cleared `waitfor_caps_pending`, does that mean `waitfor_caps` would be empty at this point?

Comment (batrick) [reply src/client/Client.cc:4286]: Yes.

File: src/client/Client.cc (hunk 1 above)
Comment (batrick) [reply src/client/Client.cc:4268]: I prefer explicit namespacing (and we generally do so everywhere).

**Comment by batrick** [comment]
jenkins test make check arm64

File: src/client/Client.cc (hunk 1 above)
Comment (dparmar18) [reply src/client/Client.cc:4268]: so should we remove the import at line 159?

Comment (batrick) [reply src/client/Client.cc:4268]: in a separate PR, sure.

**Comment by batrick** [comment]
This PR is under test in [https://tracker.ceph.com/issues/67563](https://tracker.ceph.com/issues/67563).

**Comment by batrick** [comment]
https://tracker.ceph.com/issues/67563#note-1

**Review by vshankar — APPROVED** [review]

**Pull request merged**